enclave/core/tracee.c

```diff
@@ -69,27 +77,75 @@ void oe_log_init_ecall(const char* enclave_path, uint32_t log_level)
     }
 }
 
-oe_result_t oe_log(oe_log_level_t level, const char* fmt, ...)
-{
-    oe_result_t result = OE_FAILURE;
-    oe_va_list ap;
-    int n = 0;
-    int bytes_written = 0;
-    char* message = NULL;
+void* oe_enclave_log_context = NULL;
+oe_enclave_log_callback_t oe_enclave_log_callback = NULL;
 
-    // Skip logging for non-debug-allowed enclaves
-    if (!is_enclave_debug_allowed())
+oe_result_t oe_enclave_log_set_callback(
+    void* context,
+    oe_enclave_log_callback_t callback)
+{
+    if (oe_mutex_lock(&_log_lock) == OE_OK)
     {
-        result = OE_OK;
-        goto done;
+        oe_enclave_log_context = context;
+        oe_enclave_log_callback = callback;
+        oe_mutex_unlock(&_log_lock);
+        OE_TRACE_INFO("enclave logging callback is set");
+
+        return OE_OK;
     }
 
-    // Check if this message should be skipped
-    if (level > _active_log_level)
+    return OE_UNEXPECTED;
+}
+
+oe_result_t oe_log_enclave_message(oe_log_level_t level, const char* message)
+{
+    oe_result_t result = OE_FAILURE;
+
+    // Take the log file lock.
+    if (oe_mutex_lock(&_log_lock) == OE_OK)
     {
+        if (oe_enclave_log_callback)
+        {
+            (oe_enclave_log_callback)(
+                oe_enclave_log_context,
+                level,
+                (uint64_t)oe_thread_self(),
+                message);
+        }
+        else
+        {
+            // Skip logging for non-debug-allowed enclaves
+            if (!is_enclave_debug_allowed())
```

Comment:
Unless there is a specific use case to the contrary, I would suggest that the `is_enclave_debug_allowed()` check be lifted out to 103, where it applies regardless and outside the lock.

Without a clarification helping a user navigate how to sanitise log entries, trying to handle correctly them in a non-debug context seems risky at best?

Reply:
Thank you for pointing this out! It is better to not have the user handle the enclave logs outside of debug context. Moved it back as part of basic checks, irrespective callback.